File: app/src/main/java/com/zwstudio/logicpuzzlesandroid/puzzles/tatamino/domain/TataminoGameState.java
package com.zwstudio.logicpuzzlesandroid.puzzles.tatamino.domain;

import com.rits.cloning.Cloner;
import com.zwstudio.logicpuzzlesandroid.common.domain.CellsGameState;
import com.zwstudio.logicpuzzlesandroid.common.domain.Graph;
import com.zwstudio.logicpuzzlesandroid.common.domain.GridDots;
import com.zwstudio.logicpuzzlesandroid.common.domain.GridLineObject;
import com.zwstudio.logicpuzzlesandroid.common.domain.Node;
import com.zwstudio.logicpuzzlesandroid.common.domain.Position;
import com.zwstudio.logicpuzzlesandroid.home.domain.HintState;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

import static fj.data.HashMap.fromMap;
import static fj.data.List.iterableList;

public class TataminoGameState extends CellsGameState<TataminoGame, TataminoGameMove, TataminoGameState> {
    protected Cloner cloner = new Cloner();

    public char[] objArray;
    public GridDots dots;
    public Map<Position, HintState> pos2state = new HashMap<>();

    public TataminoGameState(TataminoGame game) {
        super(game);
        objArray = new char[rows() * cols()];
        System.arraycopy(game.objArray, 0, objArray, 0, objArray.length);
        updateIsSolved();
    }

    public char get(int row, int col) {
        return objArray[row * cols() + col];
    }
    public char get(Position p) {
        return get(p.row, p.col);
    }
    public void set(int row, int col, char obj) {
        objArray[row * cols() + col] = obj;
    }
    public void set(Position p, char obj) {
        set(p.row, p.col, obj);
    }

    public boolean setObject(TataminoGameMove move) {
        Position p = move.p;
        if (!isValid(p) || get(p) == move.obj) return false;
        set(p, move.obj);
        updateIsSolved();
        return true;
    }

    public boolean switchObject(TataminoGameMove move) {
        Position p = move.p;
        if (!isValid(p) || game.get(p) != ' ') return false;
        char o = get(p);
        move.obj = o == ' ' ? '1' : o == '3' ? ' ' : (char)(o + 1);
        return setObject(move);
    }

    /*
        iOS Game: Logic Games/Puzzle Set 13/Tatamino

        Summary
        Which is a little Tatami

        Description
        1. Plays like Fillomino, in which you have to guess areas on the board
           marked by their number.
        2. In Tatamino, however, you only have areas 1, 2 and 3 tiles big.
        3. Please remember two areas of the same number size can't be touching.
    */
    private void updateIsSolved() {
        isSolved = true;
        Graph g = new Graph();
        Map<Position, Node> pos2node = new HashMap<>();
        for (int r = 0; r < rows(); r++)
            for (int c = 0; c < cols(); c++) {
                Position p = new Position(r, c);
                if (get(p) == ' ')
                    isSolved = false;
                else {
                    Node node = new Node(p.toString());
                    g.addNode(node);
                    pos2node.put(p, node);
                }
            }
        for (int r = 0; r < rows(); r++)
            for (int c = 0; c < cols(); c++) {
                Position p = new Position(r, c);
                char ch = get(p);
                if (ch == ' ') continue;
                for (Position os : TataminoGame.offset) {
                    Position p2 = p.add(os);
                    if (isValid(p2) && get(p2) == ch)
                        g.connectNode(pos2node.get(p), pos2node.get(p2));
                }
            }
        dots = cloner.deepClone(game.dots);
        while (!pos2node.isEmpty()) {
            g.setRootNode(iterableList(pos2node.values()).head());
            List<Node> nodeList = g.bfs();
            List<Position> area = fromMap(pos2node).toStream().filter(e -> nodeList.contains(e._2())).map(e -> e._1()).toJavaList();
            for (Position p : area)
                pos2node.remove(p);
            char ch = get(area.get(0));
            int n1 = area.size(), n2 = ch - '0';
            HintState s = n1 < n2 ? HintState.Normal : n1 == n2 ? HintState.Complete : HintState.Error;
            for (Position p : area) {
                pos2state.put(p, s);
                for (int i = 0; i < 4; i++) {
                    Position p2 = p.add(TataminoGame.offset[i]);
                    char ch2 = !isValid(p2) ? '.' : get(p2);
                    if (ch2 != ch && (n1 <= n2 || ch2 != ' '))
                        dots.set(p.add(TataminoGame.offset2[i]), TataminoGame.dirs[i], GridLineObject.Line);
                }
            }
            if (s != HintState.Complete) isSolved = false;
        }
    }
}

File: app/src/main/java/com/zwstudio/logicpuzzlesandroid/puzzles/tatamino/domain/TataminoGameState.kt
package com.zwstudio.logicpuzzlesandroid.puzzles.tatamino.domainimport

import com.rits.cloning.Cloner
import com.zwstudio.logicpuzzlesandroid.common.domain.*
import com.zwstudio.logicpuzzlesandroid.home.domain.HintState

class TataminoGameState(game: TataminoGame) : CellsGameState<TataminoGame, TataminoGameMove, TataminoGameState>(game) {
    var cloner = Cloner()
    var objArray = game.objArray.copyOf()
    lateinit var dots: GridDots
    var pos2state = mutableMapOf<Position, HintState>()

    operator fun get(row: Int, col: Int) = objArray[row * cols() + col]
    operator fun get(p: Position) = this[p.row, p.col]
    operator fun set(row: Int, col: Int, obj: Char) {objArray[row * cols() + col] = obj}
    operator fun set(p: Position, obj: Char) {this[p.row, p.col] = obj}

    init {
        updateIsSolved()
    }

    fun setObject(move: TataminoGameMove): Boolean {
        val p = move.p
        if (!isValid(p) || this[p] == move.obj) return false
        this[p] = move.obj
        updateIsSolved()
        return true
    }

    fun switchObject(move: TataminoGameMove): Boolean {
        val p = move.p
        if (!isValid(p) || game[p] != ' ') return false
        val o = this[p]
        move.obj = if (o == ' ') '1' else if (o == '3') ' ' else o + 1
        return setObject(move)
    }

    /*
        iOS Game: Logic Games/Puzzle Set 13/Tatamino

        Summary
        Which is a little Tatami

        Description
        1. Plays like Fillomino, in which you have to guess areas on the board
           marked by their number.
        2. In Tatamino, however, you only have areas 1, 2 and 3 tiles big.
        3. Please remember two areas of the same number size can't be touching.
    */
    private fun updateIsSolved() {
        isSolved = true
        val g = Graph()
        val pos2node = mutableMapOf<Position, Node>()
        for (r in 0 until rows())
            for (c in 0 until cols()) {
                val p = Position(r, c)
                if (this[p] == ' ')
                    isSolved = false
                else {
                    val node = Node(p.toString())
                    g.addNode(node)
                    pos2node[p] = node
                }
            }
        for (r in 0 until rows())
            for (c in 0 until cols()) {
                val p = Position(r, c)
                val ch = this[p]
                if (ch == ' ') continue
                for (os in TataminoGame.offset) {
                    val p2 = p.add(os)
                    if (isValid(p2) && this[p2] == ch)
                        g.connectNode(pos2node[p], pos2node[p2])
                }
            }
        dots = cloner.deepClone(game.dots)
        while (pos2node.isNotEmpty()) {
            g.setRootNode(pos2node.values.first())
            val nodeList = g.bfs()
            val area = pos2node.filter { nodeList.contains(it.value) }.map { it.key }
            for (p in area)
                pos2node.remove(p)
            val ch = this[area[0]]
            val n1 = area.size
            val n2 = ch - '0'
            val s = if (n1 < n2) HintState.Normal else if (n1 == n2) HintState.Complete else HintState.Error
            for (p in area) {
                pos2state[p] = s
                for (i in 0..3) {
                    val p2 = p.add(TataminoGame.offset[i])
                    val ch2 = if (!isValid(p2)) '.' else this[p2]
                    if (ch2 != ch && (n1 <= n2 || ch2 != ' '))
                        dots[p.add(TataminoGame.offset2[i]), TataminoGame.dirs[i]] = GridLineObject.Line
                }
            }
            if (s != HintState.Complete) isSolved = false
        }
    }
}